File: LeetCodeProblemsCMake/Problems/LeetCode2264Largest3SameDigitNumberInString.cpp
//
// Created by saeed on 8/14/25.
//

#include "LeetCode2264Largest3SameDigitNumberInString.h"
using namespace std;

string LeetCode2264Largest3SameDigitNumberInString::largestGoodInteger(string num) {
    string result{""};
    char largestDigitFound = '0' - 1;
    int occurance{0};
    int lastDigit{-1};
    for (int i = 0, e = static_cast<int>(num.length()); i < e; i++) {
        if (num[i] < largestDigitFound) {
            occurance = 1;
            lastDigit = num[i];
            continue;
        }
        if (num[i] != lastDigit) {
            occurance = 1;
            lastDigit = num[i];
            continue;
        }
        occurance++;
        if (occurance == 3) {
            if (num[i] > largestDigitFound) {
                largestDigitFound = num[i];
            }

            if (num[i] == '9') {
                break;
            }
        }
    }
    if (largestDigitFound >= '0') {
        result = largestDigitFound;
        result += largestDigitFound;
        result += largestDigitFound;
    }
    return result;
}

File: LeetCodeProblemsGo/LeetCode2264Largest3SameDigitNumberInString/LeetCode2264Largest3SameDigitNumberInString.go
package LeetCode2264Largest3SameDigitNumberInString

func largestGoodInteger(num string) string {
	result := ""

	largestFound := '0' - 1
	lastDigit := '0'
	occurance := 0
	for _, char := range num {
		if char < largestFound {
			lastDigit = '0' - 1
			continue
		}
		if char != lastDigit {
			occurance = 1
			lastDigit = char
			continue
		}
		occurance++
		if occurance == 3 {
			if lastDigit > largestFound {
				largestFound = lastDigit
			}
			if lastDigit == '9' {
				break
			}
		}
	}
	if largestFound < '0' {
		return ""
	}
	result = string(largestFound)
	result = result + result + result
	return result
}
